### PR TITLE
Implement finish example and output for CCB

### PR DESCRIPTION
**Pull request by jackgerrits**

### CODE REVIEW

File: vowpalwabbit/conditional_contextual_bandit.cc [hunk 1]
```diff
@@ -211,6 +212,160 @@ void learn_or_predict(ccb& data, multi_learner& base, multi_ex& examples)
...
+    std::ostringstream pred_buf;
+    pred_buf << std::setw(all.sd->col_current_predict) << std::right << std::setfill(' ') << pred_str;
+
+
+    all.sd->print_update(all.holdout_set_off, all.current_pass, label_buf.str(), pred_buf.str(), num_features,
+        all.progress_add, all.progress_arg);
+  }
+}
+
+inline float get_unbiased_cost(CCB::conditional_contexual_bandit_outcome* outcome, uint32_t action, float offset = 0.)
```
Comment (JohnLangford) [inline vowpalwabbit/conditional_contextual_bandit.cc]: Why can't we share this function with other implementations?

Comment (jackgerrits) [reply vowpalwabbit/conditional_contextual_bandit.cc]: CCB stores the cost and probability in it's own outcome structure and so we need a function that understands this. I can definitely move it to a common location though

Comment (jackgerrits) [reply vowpalwabbit/conditional_contextual_bandit.cc]: So the difference between CCB class and CB class is that CCB _can_ contain the entire pdf for that slot. So far that information has not been utilized on the reduction, so do we actually need it? If we don't then CCB can simply use the CB class

Comment (jackgerrits) [reply vowpalwabbit/conditional_contextual_bandit.cc]: @JohnLangford can you comment on the above? ^

File: vowpalwabbit/conditional_contextual_bandit.cc [hunk 2]
```diff
@@ -211,6 +212,160 @@ void learn_or_predict(ccb& data, multi_learner& base, multi_ex& examples)
...
+    return (outcome->cost - offset) / outcome->probabilities[0].score;
+  return 0.;
+}
+
+void output_example(vw& all, ccb& /*c*/, multi_ex& ec_seq)
+{
+  if (ec_seq.size() <= 0)
+    return;
+
+  std::vector<example*> decisions;
```
Comment (kumpera) [inline vowpalwabbit/conditional_contextual_bandit.cc]: Should we reserve an initial size based on the size of  `ec_seq`?

File: vowpalwabbit/conditional_contextual_bandit.cc [hunk 3]
```diff
@@ -211,6 +212,160 @@ void learn_or_predict(ccb& data, multi_learner& base, multi_ex& examples)
...
+
+  // What does it mean for not all of the decisions to be labeled? Does it become a non-labeled example at that point?
+  // Is it hold out?
+  bool labeled_example = true;
+  auto preds = ec_seq[0]->pred.decision_scores;
+  for (int i = 0; i < decisions.size(); i++)
+  {
+    if (decisions[i]->l.conditional_contextual_bandit.outcome != nullptr)
+    {
+      float l = get_unbiased_cost(decisions[i]->l.conditional_contextual_bandit.outcome, preds[i][0].action);
```
Comment (kumpera) [inline vowpalwabbit/conditional_contextual_bandit.cc]: 0 is a cool number, but a constant would be amazing.

File: vowpalwabbit/conditional_contextual_bandit.cc [hunk 4]
```diff
@@ -211,6 +212,160 @@ void learn_or_predict(ccb& data, multi_learner& base, multi_ex& examples)
...
+    }
+    ss << '\n';
+    ssize_t len = ss.str().size();
+    ssize_t t = io_buf::write_file_or_socket(f, ss.str().c_str(), (unsigned int)len);
+    if (t != len)
+      std::cerr << "write error: " << strerror(errno) << std::endl;
+  }
+}
+
+void print_update(vw& all, std::vector<example*> decisions, decision_scores_t decision_scores, size_t num_features)
```
Comment (kumpera) [inline vowpalwabbit/conditional_contextual_bandit.cc]: Pass `decisions` byref

File: vowpalwabbit/conditional_contextual_bandit.cc (hunk 2 above)
Comment (jackgerrits) [reply vowpalwabbit/conditional_contextual_bandit.cc]: We don't know how many decisions exist in the sequence and therefore may overshoot significantly. I don't think it is worth reserving

**Pull request merged**